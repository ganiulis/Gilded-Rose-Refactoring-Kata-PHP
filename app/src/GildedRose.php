<?php

declare(strict_types=1);

namespace GildedRose;

use GildedRose\Updater;

/**
 * Used for updating an array of Items.
 */
final class GildedRose
{
    private const AGED_BRIE = 'Aged Brie';
    private const BACKSTAGE_PASSES = 'Backstage passes to a TAFKAL80ETC concert';
    private const CONJURED = 'Conjured Item';
    private const SULFURAS = 'Sulfuras, Hand of Ragnaros';

    /**
     * Updated quality of selected array of Items
     *
     * @param array $items Items array to be updated
     * @return array updated Items array
     */
    public function updateItems(array $items): array
    {
        foreach ($items as $item) {
            $itemType = preg_match('/\bConjured\b/i', $item->name) ? self::CONJURED : $item->name;

            switch($itemType) {
                case self::AGED_BRIE:
                    $this->itemUpdater = new Updater\BrieUpdater;
                    break;
                case self::BACKSTAGE_PASSES:
                    $this->itemUpdater = new Updater\BackstageUpdater;
                    break;
                case self::CONJURED:
                    $this->itemUpdater = new Updater\ConjuredUpdater;
                    break;
                case self::SULFURAS:
                    $this->itemUpdater = new Updater\SulfurasUpdater;
                    break;
                default:
                    $this->itemUpdater = new Updater\DefaultUpdater;
                    break;
                }
            $this->itemUpdater->updateItem($item);
        }
        return $items;
    }
}
