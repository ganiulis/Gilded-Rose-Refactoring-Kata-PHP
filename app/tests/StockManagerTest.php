<?php 

namespace Tests\Updater;

use GildedRose\Item;
use GildedRose\StockManager;
use GildedRose\Updater;
use GildedRose\Validator\QualityValidator;
use PHPUnit\Framework\TestCase;

class StockProcesssorTest extends TestCase
{
    public function testUpdateAll(): void
    {
        $testItems = [
            new Item('foo', 4, 3),
            new Item('bar', 5, 2)
        ];

        $mockDefaultUpdater = $this->createMock(Updater\DefaultUpdater::class);

        $mockDefaultUpdater->expects($this->once())
            ->method('update')
            ->with($testItems[1])
            ->willReturn(new Item('bar', 4, 1));

        $mockUpdater = $this->createMock(Updater\BackstageUpdater::class);

        $mockUpdater->expects($this->exactly(2))
            ->method('supports')
            ->withConsecutive(
                [$testItems[0]],
                [$testItems[1]]
            )
            ->willReturnOnConsecutiveCalls(
                true,
                false
            );
        $mockUpdater->expects($this->once())
            ->method('update')
            ->with($testItems[0])
            ->willReturn(new Item('foo', 3, 2));

        $mockValidator = $this->createMock(QualityValidator::class);

        $mockValidator->expects($this->exactly(2))
            ->method('validate')
            ->withConsecutive(
                [$testItems[0]],
                [$testItems[1]]
            )
            ->willReturnOnConsecutiveCalls(
                new Item('foo', 3, 2),
                new Item('bar', 4, 1)
            );

        $manager = new StockManager(
            $mockDefaultUpdater,
            [
                $mockUpdater
            ],
            $mockValidator
        );

        $manager->updateAll($testItems);
        $manager->validateAll($testItems);
    }
}
