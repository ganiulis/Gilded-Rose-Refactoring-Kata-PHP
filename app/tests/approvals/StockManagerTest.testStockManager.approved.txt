OMGHAI!
-------- day 0 --------
name, sellIn, quality
foo, 0, 3
bar, 1, 2
zim, 2, 1
gir, 3, 0
dib, 4, -1

-------- day 1 --------
name, sellIn, quality
foo, 0, 3
bar, 1, 2
zim, 2, 1
gir, 3, 0
dib, 4, -1

-------- day 2 --------
name, sellIn, quality
foo, 0, 3
bar, 1, 2
zim, 2, 1
gir, 3, 0
dib, 4, -1

-------- day 3 --------
name, sellIn, quality
foo, 0, 3
bar, 1, 2
zim, 2, 1
gir, 3, 0
dib, 4, -1

-------- day 4 --------
name, sellIn, quality
foo, 0, 3
bar, 1, 2
zim, 2, 1
gir, 3, 0
dib, 4, -1

